<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="David Ramotowski" />

<meta name="date" content="2022-12-22" />

<title>Running the DADA2 Pipeline with 16S rRNA Sequencing Data</title>

<script src="site_libs/header-attrs-2.19/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Main Steps</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="A_Setup_and_Running_DADA2_Pipeline.html">DADA2 Pipeline</a>
</li>
<li>
  <a href="B_Analyzing_DADA2_Output_with_Phyloseq.html">Phyloseq</a>
</li>
<li>
  <a href="C_Predicting_Functional_Gene_Abundance_with_PICRUSt2_and_ALDEx2.html">PICRUSt2 and ALDEx2</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Running the DADA2 Pipeline with 16S rRNA Sequencing Data</h1>
<h4 class="author">David Ramotowski</h4>
<h4 class="date">12/22/2022</h4>

</div>


<p>Remember to replace file names and paths with your own! -Make sure all sample files are in 1 folder and in fastq.gz format -On IDAS, you should place all files into a Zip folder and upload, they will be automatically extracted</p>
<p>You will need the following R packages: -dada2, ShortRead, Biostrings, tidyverse, phyloseq, biomformat, vegan, ggbeeswarm, DESeq2, ggrepel</p>
<p>-dada2 installation code: if (!requireNamespace(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”) BiocManager::install(“dada2”)</p>
<p>-ShortRead installation code: if (!require(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”) BiocManager::install(“ShortRead”)</p>
<p>-Biostrings installation code: if (!require(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”) BiocManager::install(“Biostrings”)</p>
<p>-tidyverse installation code: install.packages(“tidyverse”)</p>
<p>-phyloseq installation code: if (!require(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”) BiocManager::install(“phyloseq”)</p>
<p>-vegan installation code: install.packages(vegan)</p>
<p>-ggbeeswarm installation code: install.packages(‘ggbeeswarm’)</p>
<p>-DESeq2 installation code: if (!require(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”) BiocManager::install(“DESeq2”)</p>
<p>-ggrepel installation code: install.packages(“ggrepel”)</p>
<p>-biomformat installation code: if (!require(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”) BiocManager::install(“biomformat”)</p>
<p>-ALDEx2 installation code: if (!require(“BiocManager”, quietly = TRUE)) install.packages(“BiocManager”)</p>
<p>BiocManager::install(“ALDEx2”)</p>
<p>-data.table installation code: install.packages(“data.table”)</p>
<p>-GUniFrac installation code: install.packages(“GUniFrac”)</p>
<p>-R.utils installation code: install.packages(“R.utils”)</p>
<p>-plyr installation code: install.packages(“plyr”)</p>
<p>-ggpubr installation code: if(!require(devtools)) install.packages(“devtools”)<br />
devtools::install_github(“kassambara/ggpubr”)</p>
<p>install.packages(“ggpubr”)</p>
<pre class="r"><code>library(dada2)
library(ShortRead)
library(Biostrings)
library(tidyverse)
library(phyloseq)
library(vegan)
library(data.table)
library(DESeq2)
library(ggrepel)
library(ggbeeswarm)
library(biomformat)
library(data.table)
library(ALDEx2)
library(GUniFrac)
library(R.utils)
library(plyr)
library(ggpubr)
#These packages need to be loaded with each new session</code></pre>
<p>You will also need to install cutadapt: (<a href="https://cutadapt.readthedocs.io/en/stable/installation.html" class="uri">https://cutadapt.readthedocs.io/en/stable/installation.html</a>)</p>
<pre class="bash"><code>python3 -m pip install --user --upgrade cutadapt
      
      ~/.local/bin/cutadapt --version</code></pre>
<p>Enter the file path for cutadapt:</p>
<pre class="r"><code>cutadapt &lt;-&quot;/home/djramotowski/.local/bin/cutadapt&quot;
    #replace my hawkid with yours ex: /home/hawkid/.local/bin/cutadapt
  system2(cutadapt, args = &quot;--version&quot;)</code></pre>
<p>For taxonomy analysis, download the Silva reference database: (<a href="https://zenodo.org/record/4587955#.YkP5F-rMKUk" class="uri">https://zenodo.org/record/4587955#.YkP5F-rMKUk</a>) Place in same folder as your samples (or you can specify the file path if it is in a different location at the trimming step)</p>
<p>-Other classifiers ex. Greengenes may be used as well</p>
<p>–Begin Analysis–</p>
<p>First, set your working directory (substitute your HawkID in place of mine):</p>
<pre class="r"><code> setwd(&quot;/home/djramotowski/DJR_Mattes_Lab/R_Script_3-27-22&quot;)
list.files() #Make sure that all fastq.gz files are in one folder</code></pre>
<p>Part 1: Removing Primers and Adapters</p>
<p>Generate paired forward and reverse read files and parse out sample names</p>
<pre class="r"><code>  fnFs &lt;-sort(list.files(pattern=&quot;_R1_001.fastq.gz&quot;, full.names = TRUE))
  fnRs &lt;-sort(list.files(pattern=&quot;_R2_001.fastq.gz&quot;, full.names = TRUE))</code></pre>
<p>Record forward and reverse primer sequences</p>
<pre class="r"><code>  FWD&lt;-&quot;GTGCCAGCMGCCGCGGTAA&quot; #Will need to change to your primers
  REV&lt;-&quot;GGACTACHVGGGTWTCTAAT&quot;</code></pre>
<p>Verify primer presence and orientation in reads</p>
<pre class="r"><code>  allOrients&lt;-function(primer) {
    require(Biostrings)
    dna &lt;- DNAString(primer)
    orients &lt;- c(Forward=dna, Complement=complement(dna),
                 Reverse=reverse(dna), RevComp=reverseComplement(dna))
    return(sapply(orients, toString))
  }
  FWD.orients &lt;- allOrients(FWD)
  REV.orients &lt;- allOrients(REV)
  FWD.orients
  REV.orients</code></pre>
<p>Remove ambiguous bases (Ns) in the sequence reads</p>
<pre class="r"><code>  fnFs.filtN &lt;- file.path(&quot;filtN&quot;, basename(fnFs))
  fnRs.filtN &lt;- file.path(&quot;filtN&quot;, basename(fnRs))
  filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0,
                multithread=TRUE) #creates an output directory named FiltN</code></pre>
<p>Check sequence quality</p>
<pre class="r"><code>  plotQualityProfile(fnFs[1:2])
  plotQualityProfile(fnRs[1:2])
    #adding qualtiy check here because sometimes quality cannot be checked
    #following primer removal [1:2] checks first two reads.</code></pre>
<p>Count number of times primers appear in forward and reverse read</p>
<pre class="r"><code>  primerHits &lt;- function(primer, fn) {
    nhits &lt;- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE) 
    return(sum(nhits&gt;0))
  }
  rbind(FWD.ForwardReads=sapply(FWD.orients, primerHits, fn=fnFs.filtN[[1]]),
        FWD.ReverseReads = sapply(FWD.orients, primerHits, fn=fnRs.filtN[[1]]),
        REV.ForwardReads = sapply(REV.orients, primerHits, fn=fnFs.filtN[[1]]),
        REV.ReverseReads = sapply(REV.orients, primerHits, fn=fnRs.filtN[[1]]))</code></pre>
<p>Remove primers using cutadapt</p>
<pre class="r"><code>  path.cut &lt;- file.path(&quot;cutadapt&quot;)
  if(!dir.exists(path.cut)) dir.create(path.cut)
  fnFs.cut &lt;- file.path(path.cut, basename(fnFs))
  fnRs.cut &lt;- file.path(path.cut, basename(fnRs))
  FWD.RC &lt;- dada2:::rc(FWD)
  REV.RC &lt;- dada2:::rc(REV)
  R1.flags &lt;- paste(&quot;-g&quot;, FWD, &quot;-a&quot;, REV.RC)
  R2.flags &lt;- paste(&quot;-G&quot;, REV, &quot;-A&quot;, FWD.RC)
  for(i in seq_along(fnFs)) {
    system2(cutadapt, args=c(R1.flags, R2.flags, &quot;-n&quot;, 2, &quot;-o&quot;, fnFs.cut[i],
                             &quot;-p&quot;, fnRs.cut[i], fnFs.filtN[i], fnRs.filtN[i]))
  }</code></pre>
<p>Sanity check for presence of primers in first cutadapt-ed sample</p>
<pre class="r"><code>  rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn=fnFs.cut[[1]]),
        FWD.ReverseReads = sapply(FWD.orients, primerHits, fn=fnRs.cut[[1]]),
        FWD.ForwardReads = sapply(REV.orients, primerHits, fn=fnFs.cut[[1]]),
        FWD.ReverseReads = sapply(REV.orients, primerHits, fn=fnRs.cut[[1]]))</code></pre>
<p>Part 2: Quality Control: Trim, Filter, Err Model, and Merge Pairs in DADA2</p>
<p>Inspecting trimmed files (replace file path here)</p>
<pre class="r"><code>  cut_path &lt;- &quot;/home/djramotowski/DJR_Mattes_Lab/R_Script_3-27-22/cutadapt&quot;
  list.files(cut_path)
  cutFs &lt;- sort(list.files(cut_path, pattern = &quot;_R1_001.fastq.gz&quot;, full.names = 
                             TRUE))
  cutRs &lt;- sort(list.files(cut_path, pattern = &quot;_R2_001.fastq.gz&quot;, full.names = 
                             TRUE))
  get.sample.name &lt;- function(fname) strsplit(basename(fname), &quot;_&quot;)[[1]][1]
  sample.names &lt;- unname(sapply(cutFs, get.sample.name))
  head(sample.names)
    #I would plot quality profiles here, but I have not gotten it to work.
      #plotQualityProfile(cutFs[1:2])
      #plotQualityProfile(cutRs[1:2])</code></pre>
<p>Filter and trim sequences</p>
<pre class="r"><code>  filtFs &lt;- file.path(cut_path, &quot;filtered&quot;, basename(cutFs))
  filtRs &lt;- file.path(cut_path, &quot;filtered&quot;, basename(cutRs))
  out &lt;- filterAndTrim(cutFs, filtFs, cutRs, filtRs, maxN = 0, maxEE = c(2, 2),
    truncQ = 2, minLen = 50, rm.phix = TRUE, compress = TRUE, 
    multithread = TRUE) #These parameters can be modified based on read quality
  head(out)
  tail(out) #Checks first and last few sequences</code></pre>
<p>Error modeling (dont’ worry if all sequences aren’t the same length)</p>
<pre class="r"><code>  errF &lt;- learnErrors(filtFs, multithread = TRUE)
  errR &lt;- learnErrors(filtRs, multithread = TRUE)
  plotErrors(errF, nominalQ = TRUE) #Plots error model
  plotErrors(errR, nominalQ = TRUE)</code></pre>
<p>Dereplicate identical reads</p>
<pre class="r"><code>  derepFs &lt;- derepFastq(filtFs, verbose = TRUE)
  derepRs &lt;- derepFastq(filtRs, verbose = TRUE)
  names(derepFs) &lt;- sample.names #Name derep-class objects by sample ID
  names(derepRs) &lt;- sample.names</code></pre>
<p>Sample inference</p>
<pre class="r"><code>  dadaFs &lt;- dada(derepFs, err = errF, multithread = TRUE)
  dadaRs &lt;- dada(derepRs, err = errR, multithread = TRUE)
  dadaFs[(1)] #can check any sequence file after quality control</code></pre>
<p>Merge forward and reverse reads</p>
<pre class="r"><code>  mergers &lt;- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)</code></pre>
<p>Construct sequence table</p>
<pre class="r"><code>  seqtab &lt;- makeSequenceTable(mergers)
  dim(seqtab)</code></pre>
<p>Remove chimeras</p>
<pre class="r"><code>  seqtab.nochim &lt;- removeBimeraDenovo(seqtab, method=&quot;consensus&quot;, 
                                      multithread=TRUE, verbose=TRUE)
  table(nchar(getSequences(seqtab.nochim)))
    #inspect distribution of sequence lengths</code></pre>
<p>Track reads through the pipeline</p>
<pre class="r"><code>  getN &lt;- function(x) sum(getUniques(x))
  track &lt;- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), 
                 sapply(mergers, getN), rowSums(seqtab.nochim))
    #For single samples, rplace sapply() with get()
  colnames(track) &lt;- c(&quot;input&quot;, &quot;filtered&quot;, &quot;denoisedF&quot;, &quot;denoisedR&quot;, &quot;merged&quot;, 
                       &quot;nonchim&quot;)
  rownames(track) &lt;- sample.names
  head(track)</code></pre>
<p>Save sequence table</p>
<pre class="r"><code>  saveRDS(seqtab, &quot;seqtab.rds&quot;)</code></pre>
<p>Part 3: Assign Taxonomy</p>
<pre class="r"><code>  taxa &lt;- assignTaxonomy(seqtab.nochim, &quot;silva_nr99_v138.1_train_set.fa.gz&quot;, 
                         multithread=TRUE)
    #Can change filepath if you don&#39;t want to copy silva into the working dir</code></pre>
<p>Inspect taxonomic assignments</p>
<pre class="r"><code>  taxa.print &lt;- taxa # Removing sequence row names for ease of visualization
  rownames(taxa.print) &lt;- NULL
  head(taxa.print)</code></pre>
<p>You can now analyze the data with packages such as Phyloseq, Vegan, etc. Good Luck!</p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
